pkg/tlscert/self_signed.go

```diff
@@ -0,0 +1,99 @@
+package tlscert
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"time"
+)
+
+// NewSelfSignedCertProvider creates a new certificate provider that creates a self-signed certificate.
+func NewSelfSignedCertProvider(commonName string, dnsNames []string, orgs []string) SelfSignedProvider {
+	return SelfSignedProvider{commonName: commonName, dnsNames: dnsNames, orgs: orgs}
+}
+
+type SelfSignedProvider struct {
+	orgs       []string
+	dnsNames   []string
+	commonName string
+}
+
+func (p SelfSignedProvider) Certificate(context.Context) (Certificate, error) {
+	// init CA config
+	ca := &x509.Certificate{
+		SerialNumber:          big.NewInt(2022),
+		Subject:               pkix.Name{Organization: p.orgs},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(1, 0, 0), // expired in 1 year
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	// generate private key for CA
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return Certificate{}, err
+	}
+
+	// create the CA certificate
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivateKey.PublicKey, caPrivateKey)
+	if err != nil {
+		return Certificate{}, err
+	}
+
+	// CA certificate with PEM encoded
+	caPEM := new(bytes.Buffer)
+	_ = pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	})
+
+	// new certificate config
+	newCert := &x509.Certificate{
+		DNSNames:     p.dnsNames,
+		SerialNumber: big.NewInt(1024),
+		Subject: pkix.Name{
+			CommonName:   p.commonName,
+			Organization: p.orgs,
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(1, 0, 0), // expired in 1 year
```

Comment:
What happens when it does expire? Should someone manually delete the secret and restart the operator pod?

Reply:
Oh, that was a can of worm. Good catch.

I added a configurable expiration time in https://github.com/grafana/rollout-operator/pull/25/commits/613d0fc4f70f9d51cc6f525294f336bbfb31575b and also added renewal of expired certificates and automatic restart after expiration date, so it would update both the secret and the webhook.